src/log_handling/RetailLogHandler.ts

```diff
@@ -0,0 +1,344 @@
+        console.debug("[RetailLogHandler] Handling CHALLENGE_MODE_START line:", line);
+
+        // It's impossible to start a keystone dungeon while another one is in progress
+        // so we'll just remove the existing one and make a new one when `CHALLENGE_MODE_START`
+        // is encountered. If any other activity is in progress, we will just exit.
+        if (this.activity) {
+            const activeChallengeMode = (this.activity.getCategory() === VideoCategory.MythicPlus);
+
+            if (activeChallengeMode) {
+                console.warn("[RetailLogHandler] A challenge mode instance is already in progress; abandoning it.")
+            } else {
+                console.error("[RetailLogHandler] Another activity in progress, can't start challenge mode.");
+                return;
+            }
+        }
+
+        const zoneName = line.arg(2);
+        const mapId = parseInt(line.arg(3), 10);
+        const hasDungeonMap = (mapId in dungeonsByMapId);
+        const hasTimersForDungeon = (mapId in dungeonTimersByMapId);
+
+        if (!hasDungeonMap || !hasTimersForDungeon) {
+            console.error(`[RetailLogHandler] Invalid/unsupported mapId for Challenge Mode dungeon: ${mapId} ('${zoneName}')`)
+        }
+
+        const startTime = line.date();
+        const zoneID = parseInt(line.arg(2), 10);
+        const level = parseInt(line.arg(4), 10);
+        const dungeonAffixes = line.arg(5).map((v: string) => parseInt(v, 10));
+
+        this.activity = new ChallengeModeDungeon(startTime, zoneID, mapId, level, dungeonAffixes);
+        const challengeModeActivity = (this.activity as ChallengeModeDungeon);
+
+        challengeModeActivity.addTimelineSegment(new ChallengeModeTimelineSegment(
+            TimelineSegmentType.Trash, this.activity.getStartDate(), 0
+        ));
+
+        this.startRecording(this.activity);
+    };
+
+    /**
+     * Handle a log line for CHALLENGE_MODE_END
+     */
+    handleChallengeModeEndLine (line: LogLine): void {
+        console.debug("[RetailLogHandler] Handling CHALLENGE_MODE_STOP line:", line);
+
+        if (!this.activity) {
+            console.error("[RetailLogHandler] Challenge mode stop with no active ChallengeModeDungeon");
+            return;
+        }
+
+        const challengeModeActivity = this.activity as ChallengeModeDungeon;
+        const endDate = line.date();
+
+        // The actual log duration of the dungeon, from which keystone upgrade
+        // levels can be calculated. This includes player death penalty. 
+        const CMDuration = Math.round(parseInt(line.arg(4), 10) / 1000);
+
+        challengeModeActivity.endChallengeMode(endDate, CMDuration);
+        this.endRecording(this.activity);
+    };
+
+    handleEncounterStartLine(line: LogLine) {
+        console.debug("[RetailLogHandler] Handling ENCOUNTER_START line:", line);
+
+        if (!this.activity) {
+            super.handleEncounterStartLine(line);
+        } else {
```

Comment:
Rewrite this to remove the else, insert a return on line 134 and outdent everything under it by one.